### PR TITLE
save and restore UI state on Android

### PR DESCRIPTION
**Pull request by t-arn**
When the orientation of an Android device is changed from portrait to landscape or vice versa, Android restarts the app and the data previously displayed in the UI is lost. To prevent this, Android has 2 callback methods: onSaveInstanceState() and onRestoreInstanceState().

This PR allows the programmer to add these two callback methods in his toga.App code.
In onSaveInstanceState(), the programmer can then save the widget data and in onRestoreInstanceState(), he can restore the data after the app has restarted and the UI has been rebuilt.

This PR needs the PR https://github.com/beeware/briefcase-android-gradle-template/pull/39

There is a demo app to test this PR:
https://bitbucket.org/TomArn/demoapp.git
Branch: android_ui_state

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by t-arn** [comment]
@freakboy3742 Please review this PR

File: src/core/toga/app.py [hunk 1]
```diff
@@ -545,6 +565,23 @@ def on_exit(self, handler):
     def add_background_task(self, handler):
         self._impl.add_background_task(handler)
 
+    def onSaveInstanceState(self, outState):
+        """Override this method in your Android toga.App to save the data of your
```
Comment (freakboy3742) [inline src/core/toga/app.py:569]: So - this method is on the public interface. The fact that the method name is an exact copy of the Android name, and you reference Android in the docstring should be a red flag.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Mechanically, this looks like it hooks up the appropriate interfaces. However, there's a few problems.

1. Firstly - a cosmetic issue: while black is the preferred code style, I'd very much prefer that you *didn't* apply black to the codebase. This makes reviewing code much more difficult, as I need to audit every cosmetic change of `'` to `"` to make sure there hasn't been any substantive change. When a PR is "2 lines of code, plus 100 lines of black changes", it imposes a lot of overhead on the review process. Code additions and modifications should follow "black" style preferences - and most editors can be configured to do this.
2. The method name you've chosen on the public interface is the Android name, using "Android spelling". `onSaveInstanceState` isn't a Pythonic name. We can't avoid using Android naming when invoking Android APIs - but on the public API, we can present a clean Pythonic API. This means, at the very least, the public interface would be `on_save_instance_state`. 
3. The docstring definitely shouldn't reference Android. iOS has a similar concept of "restore state", and in the longer term, we'll definitely need to hook up that call. We aren't going to introduce a `restore_iOS_state` method - we'll be re-using the same API entry point.
4. To that end - I'm very hesitant to add a new API hook that specifically exists to service an Android requirement without understanding how that requirement will map onto other (and especially iOS) APIs. What does "restore instance state" mean? Is this something that happens when the app is closed and reopened? What would be the interpretation on desktop platforms? This is a topic that needs to be designed, not just stuck to the side of the existing API with gaffer tape.

**Comment by freakboy3742** [comment]
There's definitely something that needs to be implemented in terms of restoring app state; however, it's a much bigger project than this PR implies; I'm not convinced this specific PR is heading in the right direction.

Closing on that basis. To progress this work, we need to have a design discussion about the "big picture" of app state storage.

**Comment by t-arn** [comment]
I'm pretty sure, for Android, there is no other way than to implement onSaveInstanceState() and onRestoreInstanceState() if you want to preserve the app state when rotating the device.

**Comment by freakboy3742** [comment]
Yes - but that's at the *implementation* layer. The API we provide on *top* of that is another matter entirely.

Using a metaphor that was (and is) very common in the Django community - when remodelling a bathroom, there is plumbing, and there is porcelain. The porcelain is the interface you actually use. The plumbing is how the work is actually done. If you don't like the porcelain, you can completely replace it without affecting the plumbing. But if you don't like the plumbing, you're usually out of luck.

`onSaveInstanceState()` et al is plumbing. It's *how* the job will be done - *on Android*. But that shouldn't have any impact on the porcelain - the interface we present to users.

To make the point even more clear: on iOS, device rotation "just works" - it's a re-render of the window layout with new geometry. "Restore state" is something that only occurs when the app has been shut down by the operating system.

So - using this code - if a user has written an Android app that implements onRestoreInstanceState() - what does the iOS app do? Should it attempt to restore instance state every time the device rotates, even though it doesn't need to? 

There is a much bigger design consideration needed here. We shouldn't just cram in an onRestoreInstanceState() because Android needs one. We need to design the whole process of restoring state, with an eye to what *all* platforms need.

**Pull request closed without merging**